metaserver/src/main/scala/scala/meta/languageserver/ctags/ScalaCtags.scala

```diff
@@ -0,0 +1,35 @@
+package scala.meta.languageserver.ctags
+
+import scala.meta._
+import org.langmeta.inputs.Input
+
+object ScalaCtags {
+  def index(input: Input.VirtualFile): CtagsIndexer = {
+    val root: Source = input.parse[Source].get
+    new Traverser with CtagsIndexer {
```

Comment:
I was reading this code and I find it very similar to what I wanted to do for symbols outline. I started implementing something similar to this to traverse the tree just for the "global" symbols, but then I saw this and I'm wondering where is this `Traverser` from and how to use it. Direct search in Scalameta repo didn't give me a definition (I'm completely lost in Scalameta's project structure 🙁 )

Reply:
@laughedelic I don’t have a computer in front of me to verify this but I believe some of the bits that you can’t tell where they are defined from scalameta are actually bits of the compiler plugin.  Scala has a standard format for compiler plugins where they hook into a global content and supply their own traversed that the compiler uses to walk the ast and pull information out.  It can appear a bit magic because it’s not fully scalameta code, but partially code for the plugin framework.  I think traverser is one of these things.

Reply:
For example, see this scala plugin demo project.  Every one I have seen folllows a similar structure.  https://github.com/cb372/scalac-plugin-basic/blob/master/src/main/scala/basic/BasicPlugin.scala#L56

Reply:
In general, I think of scalameta as being intertwined with scalac so I don’t have to be, but when trying to understand scalameta source you will end up having to dig into scalac source as well as it is translating between the two ASTs.

Reply:
@ShaneDelmore thanks a lot! This is quite enlightening. I didn't know anything about it. I'm not sure if I want to dig in the scalac's source just yet, but how can one use Scalameta without reading its sources? "_

Reply:
Hmmm....any examples of what you are trying to do?  I get by mostly by keeping quasiquotes.md open, and by parsing structures I am trying to work with and looking at their .structure output.

Reply:
@laughedelic I think `Traverser` is generated with macro annotations, it's essentially one big boilerplate pattern match on an ADT with 100s of nodes. Same for `Transformer`. The `Traverser` contract is quite simple however, override `apply`, do somthing with the tree node, optionally continue visiting children with `super.apply(tree)`.

`.collect` is for example implemented on top of `Traverser`

Reply:
scalameta doesn't convert between compiler data structures, the `scalameta` module doesn't depend on scala-compiler in fact. However, scalameta (ab)uses macro annotations to a lot of the source code is hidden inside quasiquotes to be expanded at compile time. To understand the macros it's helpful to know a bit about scalac

Reply:
@laughedelic Here's how I picked up things about meta
- https://github.com/scalameta/scalameta/blob/23de1ca839e53ad6387981bd7cced35122a2cf76/scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala
- http://scalameta.org/tutorial/#Trees
- a lot of exploration in the ammonite repl
```scala

@ import $ivy.`org.scalameta::contrib:2.0.0`, scala.meta._, contrib._
import $ivy.$                             , scala.meta._, contrib._

@ q"trait A { val x: Int }"
res1: Defn.Trait = Defn.Trait(
  List(),
  Type.Name("A"),
  List(),
  Ctor.Primary(List(), _, List()),
  Template(List(), List(), Self(_, None), List(Decl.Val(List(), List(Pat.Var(Term.Name("x"))), Type.Name("Int"))))
)
```
-
 https://astexplorer.net/#/gist/104048fb30df84e64a3e46d774f26b0f/f4b78594d5b1716f5d2821ef76e8d48b3fafc8ba

writing new docs on more detailed parts like `Transformer`/`Traverser` is on my TODO!

Reply:
@olafurpg hey! thanks a lot for explaining the things. I'm using the same "exploratory"  approach too, it's quite enough for simple uses. I just often wonder what is available in general (not to reinvent the wheel): normally if there's a generated scaladoc, it may be enough or (more often) I just check the sources and see the "global symbols outline" 😄 but with this "macrofull" code it's kind of hard (I'm also very superficially familiar with macros). Another thing that bothers me is that I don't understand how code is organized in the scalameta repo. Probably there is some development dcoumentation for it that I didn't find?

Reply:
@ShaneDelmore btw, thanks for mentioning quasiquotes.md, I didn't notice it before 👍